initialization/initialization_test.go

```diff
@@ -41,425 +46,466 @@ func (l testLogger) Debug(msg string, args ...interface{}) { l.t.Logf("\tDEBUG\t
-			assert.Less(t, prev, p)
-			prev = p
-			t.Logf("num labels written: %v\n", p)
-		}
-		s, err := (init.Started())
-		assert.NoError(t, err)
-		assert.True(t, s)
-		c, err := init.Completed()
-		assert.False(t, c)
-		assert.NoError(t, err)
-		return nil
-	})
-	err = init.Initialize()
-	r.Equal(ErrStopped, err)
-	r.NoError(eg.Wait())
-
-	// Continue the initialization to completion.
-	eg.Go(assertNumLabelsWrittenChan(init, t))
-	r.NoError(init.Initialize())
-	r.NoError(eg.Wait())
-
-	// Cleanup.
-	r.NoError(init.Reset())
-}
 
-func Test_SessionNumLabelsWrittenChan_Racefree(t *testing.T) {
-	r := require.New(t)
-	log := testLogger{t: t}
+	cfg := config.DefaultConfig()
+	cfg.LabelsPerUnit = 1 << 12
 
-	cfg, opts := getTestConfig(t)
+	opts := config.DefaultInitOpts()
+	opts.DataDir = t.TempDir()
 	opts.NumUnits = 10
-	commitment := make([]byte, 32)
+	opts.NumFiles = 5
+	opts.ComputeProviderID = CPUProviderID()
 
-	init, err := NewInitializer(cfg, opts, commitment)
+	init, err := NewInitializer(
+		WithCommitment(make([]byte, 32)),
+		WithConfig(cfg),
+		WithInitOpts(opts),
+		WithLogger(testLogger{t: t}),
+	)
 	r.NoError(err)
-	init.SetLogger(log)
-
-	eg, _ := errgroup.WithContext(context.Background())
-
-	// call SessionNumLabelsWrittenChan() during initialization
-	eg.Go(func() error {
-		ch := make([]<-chan uint64, 0, 10)
-
-		for i := 0; i < 10; i++ {
-			time.Sleep(600 * time.Millisecond)
-			ch = append(ch, init.SessionNumLabelsWrittenChan())
-		}
-
-		if ch[0] != nil {
-			return errors.New("channel should be nil")
-		}
-
-		for _, c := range ch[1:] {
-			<-c
-		}
-		return nil
-	})
-	eg.Go(func() error {
-		time.Sleep(1 * time.Second)
-		return init.Initialize()
-	})
-	r.NoError(eg.Wait())
-
-	// call SessionNumLabelsWrittenChan() after initialization
-	eg.Go(func() error {
-		ch := make([]<-chan uint64, 0, 10)
-
-		for i := 0; i < 10; i++ {
-			ch = append(ch, init.SessionNumLabelsWrittenChan())
-		}
 
-		for _, c := range ch {
-			select {
-			case <-c:
-			default:
-				return errors.New("channel should be closed")
+	// Start initialization and stop it after it has written some labels
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		var eg errgroup.Group
+		eg.Go(func() error {
+			timer := time.NewTicker(50 * time.Millisecond)
+			defer timer.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-timer.C:
+					if init.SessionNumLabelsWritten() > 0 {
+						cancel()
+						return nil
+					}
+				}
 			}
-		}
-		return nil
-	})
-	r.NoError(eg.Wait())
+		})
+		r.ErrorIs(init.Initialize(ctx), context.Canceled)
+		eg.Wait()
+
+		status := init.Status()
+		assert.Equal(t, StatusStarted, status)
+	}
 
-	// Cleanup.
-	r.NoError(init.Reset())
+	// Continue the initialization to completion.
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		var eg errgroup.Group
+		eg.Go(assertNumLabelsWritten(ctx, t, init))
+		r.NoError(init.Initialize(ctx))
```

Comment:
That was my original intention; but with the current implementation I see no drawback to having `Initialize` being able to be called twice. The second call just returns (almost) immediately if initialisation previously already completed or continues where the previous call left off.

Initially I had the plan to make `Initializer::Initialize()` just a function that receives all the arguments of `NewInitializer` and a `context` that can be cancelled to abort PoST initialization. But this would require a bigger refactoring:

- `Status()` could be completely removed: the caller knows the status, either they didn't call `Initialize()` yet, they called it and it and it hasn't returned yet, they called it and it returned with an error or it completed successfully.
- `Reset()` could be its own function that just receives the `DataDir` as parameter.
- `NumLabelsWritten()` is a bit more involved but could also be a function that only requires `DataDir` as a parameter.